lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -375,6 +425,62 @@ def jobs(self):
 
         return scrape_jobs
 
+    def alerts(self):
+        """Fetch alerts for all relations.
+
+        A Prometheus alert rules file consists of a list of "groups". Each
+        group consists of a list of alerts (`rules`) that are sequentially
+        executed. This method returns all the alert rules provided by each
+        related consumer charm. These rules may be used to generate a
+        separate alert rules file for each relation since the returned list
+        of alert groups are indexed by relation ID. Also for each relation ID
+        associated scrape metadata such as Juju model, UUID and application
+        name are provided so the a unique name may be generated for the rules
+        file. For each relation the structure of data returned is a dictionary
+        with four keys
+
+        - groups
+        - model
+        - model_uuid
+        - application
+
+        The value of the `groups` key is such that it may be used to generate
+        a Prometheus alert rules file directly using `yaml.dump` but the
+        `groups` key itself must be included as this is required by Prometheus,
+        for example as in `yaml.dump({"groups": alert["groups"]})`.
+
+        Currently the consumer charm only accepts a list of rules and these
+        rules are all placed into a single group, even though Prometheus itself
+        allows for multiple groups within a single alert rules file.
+
+        Returns:
+            a dictionary of alert rule groups and associated scrape
+            metadata indexed by relation ID.
+        """
+        alerts = {}
+        for relation in self._charm.model.relations[self._relation_name]:
+            if len(relation.units) == 0:
+                continue
+
+            alert_rules = json.loads(
+                relation.data[relation.app].get("alert_rules", "{}")
+            )
+            scrape_metadata = json.loads(
+                relation.data[relation.app].get("scrape_metadata", "{}")
+            )
+            if not scrape_metadata:
+                continue
+
+            if alert_rules:
+                alerts[relation.id] = {
+                    "groups": alert_rules["groups"],
+                    "model": scrape_metadata["model"],
+                    "model_uuid": scrape_metadata["model_uuid"][:7],
+                    "application": scrape_metadata["application"],
+                }
```

Comment:
```suggestion
            if relation.units:
                # Load alert rules and scrape data from relation data bag
                alert_rules = json.loads(relation.data[relation.app].get("alert_rules", "{}"))
                scrape_metadata = json.loads(relation.data[relation.app].get("scrape_metadata", "{}"))
                
                if scrape_metadata and alert_rules:
                    alerts[relation.id] = {
                        "groups": alert_rules["groups"],
                        "model": scrape_metadata["model"],
                        "model_uuid": scrape_metadata["model_uuid"][:7],
                        "application": scrape_metadata["application"],
                    }
```

This is quite a lot shorter, and imo easier to read. One thing to check is if the top conditional is valid (i.e. `if relation.units` has the desired effect)

Reply:
Also subject to line breaks/length limits ;)

Reply:
Updated...

Reply:
Hmm :) . Still does not feel right. What if `alert_rules` exist but not `scrape_metadata` or vice versa. Seems like in this case an incomplete alert will be returned in the list of alerts. I think your previous edit was on the right track in using a conditional that was the conjunction of the two. 

But let me give it another thought. I will see if walrus operator and conjunction in the conditional shortens it. But I still prefer using negative conditionals to abort processing any alert if any one of its requirements are not met. I am inclined to convert even the `if alert_rules` into a negative conditional. 

With regard to semantics of `relation.units` and its behaviour in boolean contexts, it would save time if these things were explicitly documented thereby avoiding the need to dig into Operator Framework code base to figure it out.

Reply:
(And BTW, the behaviour you describe above is the same in the initial implementation...)

Reply:
Code has be refactored to use a conjunction as the conditional. Please check last two commits. I still retain negative conditional on `relation.units` to avoid an unnecessary extra level of nesting. But most importantly invalid rules errors are now logged and will not raise exceptions.

I will add unit tests and then request re-review.